README.md
# NAME

signatures - Subroutine signatures with no source filter

# VERSION

version 0.09

# SYNOPSIS

    use signatures;

    sub foo ($bar, $baz) {
        return $bar + $baz;
    }

# DESCRIPTION

With this module, we can specify subroutine signatures and have variables
automatically defined within the subroutine.

For example, you can write

    sub square ($num) {
        return $num * $num;
    }

and it will be automatically turned into the following at compile time:

    sub square {
        my ($num) = @_;
        return $num * $num;
    }

Note that, although the syntax is very similar, the signatures provided by this
module are not to be confused with the prototypes described in [perlsub](https://metacpan.org/pod/perlsub). All
this module does is extracting items of @\_ and assigning them to the variables
in the parameter list. No argument validation is done at runtime.

The signature definition needs to be on a single line only.

If you want to combine sub signatures with regular prototypes a `proto`
attribute exists:

    sub foo ($bar, $baz) : proto($$) { ... }

# METHODS

If you want subroutine signatures doing something that this module doesn't
provide, like argument validation, typechecking and similar, you can subclass
it and override the following methods.

## proto\_unwrap ($prototype)

Turns the extracted `$prototype` into code.

The default implementation returns `my (${prototype}) = @_;` or an empty
string, if no prototype is given.

## inject ($offset, $code)

Inserts a `$code` string into the line perl currently parses at the given
`$offset`. This is only called by the `callback` method.

## callback ($offset, $prototype)

This gets called as soon as a sub definition with a prototype is
encountered. Arguments are the `$offset` within the current line perl
is parsing and extracted `$prototype`.

The default implementation calls `proto_unwrap` with the prototype and passes
the returned value and the offset to `inject`.

# BUGS

- prototypes aren't checked for validity yet

    You won't get a warning for invalid prototypes using the `proto` attribute,
    like you normally would with warnings enabled.

- you shouldn't alter $SIG{\_\_WARN\_\_} at compile time

    After this module is loaded you shouldn't make any changes to `$SIG{__WARN__}`
    during compile time. Changing it before the module is loaded or at runtime is
    fine.

# SEE ALSO

[Method::Signatures](https://metacpan.org/pod/Method::Signatures)

[MooseX::Method::Signatures](https://metacpan.org/pod/MooseX::Method::Signatures)

[Sub::Signatures](https://metacpan.org/pod/Sub::Signatures)

[Attribute::Signature](https://metacpan.org/pod/Attribute::Signature)

[Perl6::Subs](https://metacpan.org/pod/Perl6::Subs)

[Perl6::Parameters](https://metacpan.org/pod/Perl6::Parameters)

# THANKS

Moritz Lenz and Steffen Schwigon for documentation review and
improvement.

# AUTHOR

Florian Ragwitz <rafl@debian.org>

# COPYRIGHT AND LICENSE

This software is copyright (c) 2008 by Florian Ragwitz.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

# CONTRIBUTORS

- Alex Kapranoff <alex@kapranoff.ru>
- Alexandr Ciornii <alexchorny@gmail.com>
- Karen Etheridge <ether@cpan.org>
- Peter Martini <PeterCMartini@GMail.com>
- Steffen Schwigon <ss5@renormalist.net>


Changes
Revision history for signatures

{{$NEXT}}

0.14      2019-05-03 05:39:41Z
  - properly clear PL_lex_stuff (perl RT#134058, Dave Mitchell)

0.13      2015-07-28 06:35:39Z
  - bump minimum version of B::Hooks::Parser for ::Install::Files

0.12      2015-05-13 01:23:05Z
  - fix parsing of attribute handlers (GH #1)

0.11      2015-01-03 20:35:48Z
  - re-release to fix problematic $VERSION declaration (RT#101095)

0.10      2014-12-22 08:14:23Z
  - fix compile errors on perl 5.21.6 (RT#101076, Father Chrysostomos)

0.09      2014-05-25 16:17:38Z
  - bump required version of B::Hooks::Parser for MSWin32

0.08      2014-05-24 05:07:22Z
  - Patch for perl > 5.19.2 by Peter Martini (RT#95181)

0.07      2012-05-19 10:46:23Z
  - Patch for perl > 5.13.7 by Colin Watson
    See https://rt.cpan.org/Public/Bug/Display.html?id=66997

0.06      2009-07-31 04:34:56Z
  - Require ExtUtils::Depends 0.302.

0.05      2009-04-19 17:21:35Z
  - Depend on B::Hooks::EndOfScope 0.08 so exceptions
    during compile time that trigger a destructor don't
    cause a panic on 5.10 anymore.

0.04      2009-04-19 10:00:18Z
  - Skip a test for syntax errors in string evals that
    sometimes makes perl crash.
  - Add test case for something that used to blow up.
  - Depend on latest B::Hooks::OP::Check.
  - Add repository and bugtracker urls to META.yml.

0.03      2008-12-09 13:35:24Z
  - Suppress a new prototype warning from blead perl.

0.02      2008-11-11 14:26:33Z
  - Port to B::Hooks::OP::PPAddr.

0.01      2008-11-05 08:58:17Z
  - Initial release.


lib/signatures.pm
use strict;
use warnings;
package signatures;
# ABSTRACT: Subroutine signatures with no source filter

our $VERSION = '0.15';

use XSLoader;
use B::Hooks::Parser 0.12;
use B::Hooks::OP::Check 0.17;
use B::Hooks::OP::PPAddr 0.03;
use B::Hooks::EndOfScope 0.08 ();

XSLoader::load(
    __PACKAGE__,
    $VERSION,
);

{
    my %pkgs;

    sub import {
        my ($class) = @_;
        my $caller = caller();
        $pkgs{$caller} = $class->setup_for($caller);
        return;
    }

    sub unimport {
        my ($class) = @_;
        my $caller = caller();
        $class->teardown_for(delete $pkgs{$caller});
        return;
    }
}

sub setup_for {
    my ($class, $caller) = @_;
    my $ret = $class->setup($caller);

    $^H{"${class}::enabled"} = 1;

    my $old_warn = $SIG{__WARN__};
    $SIG{__WARN__} = sub {
        if ($_[0] !~ /^(?:(?:Illegal character in prototype)|(?:Prototype after '.')) for /) {
            $old_warn ? $old_warn->(@_) : warn @_;
        }
    };

    my $unregister;
    {
        my $called = 0;
        $unregister = sub {
            return if $called++;
            $class->teardown_for([$ret, $unregister]);
            $SIG{__WARN__} = $old_warn;
        };
    }

    &B::Hooks::EndOfScope::on_scope_end($unregister);

    return [$ret, $unregister];
}

sub teardown_for {
    my ($class, $data) = @_;
    delete $^H{"${class}::enabled"};
    $class->teardown($data->[0]);
    $data->[1]->();
    return;
}

sub callback {
    my ($class, $offset, $proto) = @_;
    my $inject = $class->proto_unwrap($proto);
    $class->inject($offset, $inject);
    return;
}

sub proto_unwrap {
    my ($class, $proto) = @_;
    return '' unless length $proto;
    return "my ($proto) = \@_;";
}

sub inject {
    my ($class, $offset, $inject) = @_;
    my $linestr = B::Hooks::Parser::get_linestr();
    substr($linestr, $offset + 1, 0) = $inject;
    B::Hooks::Parser::set_linestr($linestr);
    return;
}

1;

__END__

=head1 SYNOPSIS

    use signatures;

    sub foo ($bar, $baz) {
        return $bar + $baz;
    }

=head1 DESCRIPTION

With this module, we can specify subroutine signatures and have variables
automatically defined within the subroutine.

For example, you can write

    sub square ($num) {
        return $num * $num;
    }

and it will be automatically turned into the following at compile time:

    sub square {
        my ($num) = @_;
        return $num * $num;
    }

Note that, although the syntax is very similar, the signatures provided by this
module are not to be confused with the prototypes described in L<perlsub>. All
this module does is extracting items of @_ and assigning them to the variables
in the parameter list. No argument validation is done at runtime.

The signature definition needs to be on a single line only.

If you want to combine sub signatures with regular prototypes a C<proto>
attribute exists:

    sub foo ($bar, $baz) : proto($$) { ... }

=head1 METHODS

If you want subroutine signatures doing something that this module doesn't
provide, like argument validation, typechecking and similar, you can subclass
it and override the following methods.

=head2 proto_unwrap ($prototype)

Turns the extracted C<$prototype> into code.

The default implementation returns C<< my (${prototype}) = @_; >> or an empty
string, if no prototype is given.

=head2 inject ($offset, $code)

Inserts a C<$code> string into the line perl currently parses at the given
C<$offset>. This is only called by the C<callback> method.

=head2 callback ($offset, $prototype)

This gets called as soon as a sub definition with a prototype is
encountered. Arguments are the C<$offset> within the current line perl
is parsing and extracted C<$prototype>.

The default implementation calls C<proto_unwrap> with the prototype and passes
the returned value and the offset to C<inject>.

=for Pod::Coverage setup setup_for teardown teardown_for

=head1 LIMITATIONS

=over 4

=item prototypes aren't checked for validity yet

You won't get a warning for invalid prototypes using the C<proto> attribute,
like you normally would with warnings enabled.

=item you shouldn't alter $SIG{__WARN__} at compile time

After this module is loaded you shouldn't make any changes to C<$SIG{__WARN__}>
during compile time. Changing it before the module is loaded or at runtime is
fine.

=back

=head1 SEE ALSO

=for :list
* L<Method::Signatures>
* L<MooseX::Method::Signatures>
* L<Sub::Signatures>
* L<Attribute::Signature>
* L<Perl6::Subs>
* L<Perl6::Parameters>

=head1 THANKS

=for stopwords Moritz Lenz Steffen Schwigon

Moritz Lenz and Steffen Schwigon for documentation review and
improvement.

=cut


lib/signatures/API.pm
use strict;
use warnings;

package signatures::API;

1;
__END__
=pod

=head1 NAME

signatures::API - underlying implementation for L<signatures>

=head1 METHODS

If you want subroutine signatures doing something that this module doesn't
provide, like argument validation, typechecking and similar, you can subclass
it and override the following methods.

=head2 proto_unwrap ($prototype)

Turns the extracted C<$prototype> into code.

The default implementation returns C<< my (${prototype}) = @_; >> or an empty
string, if no prototype is given.

=head2 inject ($offset, $code)

Inserts a C<$code> string into the line perl currently parses at the given
C<$offset>. This is only called by the C<callback> method.

=head2 callback ($offset, $prototype)

This gets called as soon as a sub definition with a prototype is
encountered. Arguments are the C<$offset> within the current line perl
is parsing and extracted C<$prototype>.

The default implementation calls C<proto_unwrap> with the prototype and passes
the returned value and the offset to C<inject>.

=head1 BUGS

=over 4

=item prototypes aren't checked for validity yet

You won't get a warning for invalid prototypes using the C<proto> attribute,
like you normally would with warnings enabled.

=item you shouldn't alter $SIG{__WARN__} at compile time

After this module is loaded you shouldn't make any changes to C<$SIG{__WARN__}>
during compile time. Changing it before the module is loaded or at runtime is
fine.

=back

=cut


inc/ExtUtilsDepends.pm
use strict;
use warnings;
package # hide from PAUSE
    inc::ExtUtilsDepends;

use Moose;
with
    'Dist::Zilla::Role::PrereqSource',
    'Dist::Zilla::Role::InstallTool',
;
use List::Util 'first';

sub register_prereqs {
    my $self = shift;

    $self->zilla->register_prereqs(
        {
            phase => $_,
            type  => 'requires',
        },
        # minimum version that works on Win32+gcc
        'ExtUtils::Depends' => '0.302',
        'B::Hooks::OP::Check' => '0.17',
        # minimum version that depends on ExtUtils::Depends 0.302
        'B::Hooks::OP::PPAddr' => '0.03',
        # minimum version that depends on ExtUtils::Depends 0.302
        'B::Hooks::Parser' => '0.12',
    # this is a workaround for a bug in [MakeMaker], where configure requires
    # prereqs are not added to the fallback hash
    ) foreach qw(configure build);
}

sub setup_installer {
    my $self = shift;

    my $file = first { $_->name eq 'Makefile.PL' } @{$self->zilla->files};
    $self->log_fatal('No Makefile.PL found!') if not $file;

    my $extra_content = <<'CONTENT';
use ExtUtils::Depends 0.302;
my $pkg = ExtUtils::Depends->new(
    'signatures',
    'B::Hooks::OP::Check',
    'B::Hooks::OP::PPAddr',
    'B::Hooks::Parser',
);
$pkg->add_xs('signatures.xs');
$pkg->add_pm('lib/signatures.pm' => '$(INST_LIB)/signatures.pm');
%WriteMakefileArgs = ( %WriteMakefileArgs, $pkg->get_makefile_vars );
CONTENT

    my $content = $file->content;
    $self->log_debug('Inserting extra content into Makefile.PL...');

    # this is a vicious hack -- maybe someday [MakeMaker::Awesome] will give
    # us a hook to add this content into the right spot.
    $self->log_fatal('failed to find position in Makefile.PL to munge!')
        if $content !~ m'^my %WriteMakefileArgs = \(\n(?:[^;]+)^\);$'mg;

    my $pos = pos($content);

    $content = substr($content, 0, $pos)
        . "\n\n"
        . $extra_content
        . "\n" . substr($content, $pos, -1);

    $file->content($content);
    return;
}

1;
